### PR TITLE
CDAP-17595 Added support to upgrade the pipeline config for replicator app.

### PR DESCRIPTION
**Pull request by sagarkapare**
JIRA: https://cdap.atlassian.net/browse/CDAP-17595

### CODE REVIEW

File: delta-proto/src/main/java/io/cdap/delta/proto/DeltaConfig.java [hunk 1]
```diff
@@ -211,6 +212,24 @@ public int hashCode() {
...
   }
 
+  /**
+   * Updates current DeltaConfig by running update actions provided in context such as upgrading plugin artifact
+   * versions.
+   *
+   * @param upgradeContext Context for performing update for current delta config.
+   * @return a new (updated) delta config after performing update operations.
+   */
+  public DeltaConfig updateBatchConfig(ApplicationUpdateContext upgradeContext)
```
Comment (albertshau) [inline delta-proto/src/main/java/io/cdap/delta/proto/DeltaConfig.java]: nit: updateBatchConfig -> updateConfig

File: delta-proto/src/main/java/io/cdap/delta/proto/Stage.java [hunk 2]
```diff
@@ -38,6 +49,113 @@ public Plugin getPlugin() {
...
+  /**
+   * Updates stage by performing update action logic provided in context.
+   * Current relevant update actions for stages are:
+   *  1. UPGRADE_ARTIFACT: Upgrades plugin artifact by finding the latest version of plugin to use.
+   *
+   * @param updateContext Context to use for updating stage.
+   * @return new (updated) Stage.
+   */
+  public Stage updateStage(ApplicationUpdateContext updateContext) throws Exception {
+    for (ApplicationConfigUpdateAction updateAction: updateContext.getUpdateActions()) {
```
Comment (albertshau) [inline delta-proto/src/main/java/io/cdap/delta/proto/Stage.java]: style: add space before `:`

File: delta-proto/src/main/java/io/cdap/delta/proto/Stage.java [hunk 3]
```diff
@@ -38,6 +49,113 @@ public Plugin getPlugin() {
...
+
+    if (currentVersionRange.isExactVersion()) {
+      if (currentVersionRange.getLower().compareTo(newPlugin.getVersion()) < 0) {
+        // Current version is a fixed version and new version is higher than current.
+        return newPlugin.getVersion().getVersion();
+      }
+      return null;
+    }
+
+    if (!currentVersionRange.isExactVersion()) {
```
Comment (albertshau) [inline delta-proto/src/main/java/io/cdap/delta/proto/Stage.java]: seems like this will always be true at this point

**Review by albertshau — APPROVED** [review]
minor comments.

Is this mostly copy-pasted from the pipeline upgrade code? If so, can you open a tech debt jira to consolidate the logic? I would imagine this artifact bumping logic will be common across all apps.

File: delta-proto/src/main/java/io/cdap/delta/proto/Stage.java (hunk 2 above)
Comment (sagarkapare) [reply delta-proto/src/main/java/io/cdap/delta/proto/Stage.java]: fixed.

File: delta-proto/src/main/java/io/cdap/delta/proto/Stage.java (hunk 3 above)
Comment (sagarkapare) [reply delta-proto/src/main/java/io/cdap/delta/proto/Stage.java]: good catch! fixed.

File: delta-proto/src/main/java/io/cdap/delta/proto/DeltaConfig.java (hunk 1 above)
Comment (sagarkapare) [reply delta-proto/src/main/java/io/cdap/delta/proto/DeltaConfig.java]: renamed.

**Comment by sagarkapare** [comment]
Filed - https://cdap.atlassian.net/browse/CDAP-17652 for CDAP to support plugin upgrade.

**Review by albertshau — APPROVED** [review]

**Comment by sagarkapare** [comment]
Build is green - https://builds.cask.co/browse/DEL-BT117-5. Squashing and merging.

**Pull request merged**